### PR TITLE
Ruler: remote rule evaluation

### PR DESCRIPTION
**Pull request by dannykopping**
**What this PR does / why we need it**:
Adds the ability to evaluate recording & alerting rules against a given `query-frontend`, allowing these queries to be executed with all the parallelisation & optimisation that regular adhoc queries have. This is important because with `local` evaluation all queries are single-threaded, and rules that evaluate a large range/volume of data may timeout or OOM the `ruler` itself, leading to missed metrics or alerts.

When `remote` evaluation mode is enabled, the `ruler` effectively just becomes a gRPC client for the `query-frontend`, which will dramatically improve the reliability of the `ruler` and also drastically reduce its resource requirements.

**Which issue(s) this PR fixes**:
This PR implements the feature discussed in https://github.com/grafana/loki/pull/8129 (**LID 0002: Remote Rule Evaluation**).

**Special notes for your reviewer**:
This PR is the first of two; in this PR I have just implemented the happy path. In a subsequent PR I will address all the unhappy paths and add some configurability to this feature. I didn't want to complicate this (already large) PR with all of those failure cases, as a couple of these will require some deliberation. That subsequent PR will also contain documentation.

Also, I added a couple drive-by improvements along the way:

- refactored our integration tests a little so the ruler config could be controlled by tests in a cleaner way
  - also added the ability to provide extra config fragments which are merged into the main Loki config; this makes tests quite clean and clear now I think
- added a `loki-backend` component to our `production/docker` docker-compose setup, as well as configured the ruler and added a couple sample rules
  - also added remote-writing of metrics to the `prometheus` instance

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/querier/queryrange/codec.go [hunk 1]
```diff
@@ -805,6 +817,19 @@ func toProtoVector(v loghttp.Vector) []queryrangebase.SampleStream {
 	return res
 }
 
+func toProtoScalar(v loghttp.Scalar) []queryrangebase.SampleStream {
+	res := make([]queryrangebase.SampleStream, 0, 1)
+
+	res = append(res, queryrangebase.SampleStream{
+		Samples: []logproto.LegacySample{{
+			Value:       float64(v.Value),
+			TimestampMs: v.Timestamp.UnixNano() / 1e6,
```
Comment (chaudum) [inline pkg/querier/queryrange/codec.go:826]: nit: There is a `v.Timestamp.UnixMilli()` function for that.

File: integration/client/client.go [hunk 2]
```diff
@@ -434,7 +434,7 @@ func (c *Client) GetRules(ctx context.Context) (*RulesResponse, error) {
 	resp := RulesResponse{}
 	err = json.Unmarshal(buf, &resp)
 	if err != nil {
-		return nil, fmt.Errorf("error parsing response data: %w", err)
+		return nil, fmt.Errorf("error parsing response data %q: %w", buf, err)
```
Comment (kavirajk) [inline integration/client/client.go:437]: Q: given this buffer contains results of `GetRules`. can this be huge enough to cause trouble? (also can it make too verbose error message?)

File: pkg/loki/loki.go [hunk 3]
```diff
@@ -356,6 +356,7 @@ type Loki struct {
 	tableManager             *index.TableManager
 	frontend                 Frontend
 	ruler                    *base_ruler.Ruler
+	ruleEvaluator            ruler.Evaluator
```
Comment (kavirajk) [inline pkg/loki/loki.go:359]: Q: `ruleEvaluator` runs as a part of `ruler` component/process correct?

File: pkg/querier/queryrange/codec.go [hunk 4]
```diff
@@ -521,6 +521,18 @@ func (Codec) DecodeResponse(ctx context.Context, r *http.Response, req queryrang
 				},
 				Statistics: resp.Data.Statistics,
 			}, nil
+		case loghttp.ResultTypeScalar:
```
[the remark this answers is not in the document]
Comment (kavirajk) [reply pkg/querier/queryrange/codec.go:524]: Nice!. Can we lock this in test as well if possible?

**Review by kavirajk — APPROVED** [review]
Looks super nice :+1:

File: pkg/ruler/evaluator_remote.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+// Query performs a query for the given time.
+func (q *remoteQuerier) Query(ctx context.Context, qs string, t time.Time) (*logqlmodel.Result, error) {
+	logger, ctx := spanlogger.NewWithLogger(ctx, q.logger, "ruler.remoteEvaluation.Query")
+	defer logger.Span.Finish()
+
+	res, err := q.query(ctx, qs, t, logger)
+	if err != nil {
+		return nil, err
+	}
+	return res, err
```
Comment (chaudum) [inline pkg/ruler/evaluator_remote.go:147]: nit: should be identical to:

```suggestion
	return q.query(ctx, qs, t, logger)
```

**Review by chaudum — COMMENTED** [review]
LGTM :rocket: 
I added a few comments of minor importance, but don't want to block merging.

File: integration/util/merger.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
+package util
```
[the remark this answers is not in the document]
Comment (chaudum) [reply integration/util/merger.go:1]: Would have been nice if that change would be a separate PR (would reduce size of this one)

[the remark this answers is not in the document]
Comment (dannykopping) [reply integration/util/merger.go:1]: I needed these changes to make the integration tests work; but yeah I suppose I could've left the integration tests for the second PR as well. I didn't know from the beginning I would need a subsequent PR...

File: pkg/querier/queryrange/codec.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dannykopping) [reply pkg/querier/queryrange/codec.go:524]: Yup, in the subsequent PR

File: pkg/loki/loki.go (hunk 3 above)
Comment (dannykopping) [reply pkg/loki/loki.go:359]: Correct

File: pkg/ruler/evaluator_remote.go (hunk 5 above)
Comment (dannykopping) [reply pkg/ruler/evaluator_remote.go:147]: You're right, I think I had some error wrapping there before but removed it.

File: pkg/querier/queryrange/codec.go (hunk 1 above)
Comment (dannykopping) [reply pkg/querier/queryrange/codec.go:826]: Ah, right, this is a `prometheus/common/model.Time` so there's no UnixMilli() available

**Comment by weizhengde** [comment]
when this feature will be released

**Comment by dannykopping** [comment]
We have a 2.9 release coming soon which will include this

**Pull request merged**